### PR TITLE
(PUP-10238) Change default value of strict_hostname_checking to true

### PR DESCRIPTION
**Pull request by justinstoller**
Previously our default value of strict_hostname_checking was false which
allowed matching dotted segments of a nodes certname (its CN in its
certificate) as well as the segments of its fqdn fact, or hostname +
domain fact.

This was for compatibility when fact based classification within a
site.pp was a more common pattern and node declarations were much less
powerful than they are now.

With the ability to use regular expressions in a node declaration the
auto segmenting is no longer needed and with the ability to use facts
directly, to use fact interpetation in hiera lookups, or create a custom
external node classifier the injecting of facts into the nodes "name" is
unneeded.

The desire is to remove the setting completely in Puppet 7, while
leaving it in 6 so those that depend on this behavior have time to
re-write their site.pps to the newer styles.

strict_hostname_checking setting is not marked deprecated completely
because it will cause deprecation notices on setting access, which
happens as part of normal compilation for now. However it does mark
"node_name" setting as deprecated completely because it is now only
referenced in code that by default will not run (and will only run if
users change strict_hostname_checking back to false).

### CODE REVIEW

File: lib/puppet/defaults.rb [hunk 1]
```diff
@@ -1297,13 +1297,23 @@ def self.initialize_default_settings!(settings)
         overridden by more specific settings (see `ca_port`, `report_port`).",
       },
       :node_name => {
-        :default    => "cert",
+        :default    => 'cert',
+        :type       => :enum,
+        :values     => ['cert', 'facter'],
+        :deprecated => :completely,
+        :hook       => proc { |val|
+          if value != 'cert'
```
Comment (mwaggett) [inline lib/puppet/defaults.rb]: should this be `val` instead of `value`?

File: lib/puppet/defaults.rb [hunk 2]
```diff
@@ -1297,13 +1297,23 @@ def self.initialize_default_settings!(settings)
...
+            Puppet.deprecation_warning("The node_name setting is deprecated and will be removed in a future release.")
+          end
+        },
         :desc       => "How the puppet master determines the client's identity
         and sets the 'hostname', 'fqdn' and 'domain' facts for use in the manifest,
         in particular for determining which 'node' statement applies to the client.
         Possible values are 'cert' (use the subject's CN in the client's
         certificate) and 'facter' (use the hostname that the client
-        reported in its facts)",
+        reported in its facts)
```
Comment (mwaggett) [inline lib/puppet/defaults.rb]: nitpick, but while you're in here, I think this should end in a `.`

File: spec/unit/node_spec.rb [hunk 3]
```diff
@@ -445,25 +447,30 @@ def from_json(json)
 
   describe "and :node_name is set to 'cert'" do
     before do
-      Puppet[:strict_hostname_checking] = false
       Puppet[:node_name] = "cert"
+      @node = Puppet::Node.new("foo.domain.com",
```
Comment (mwaggett) [inline spec/unit/node_spec.rb]: Why is this needed now? Looks like the same as the `@node` above?

File: spec/unit/node_spec.rb [hunk 4]
```diff
@@ -445,25 +447,30 @@ def from_json(json)
...
 
     it "uses the passed-in key as the first value" do
       expect(@node.names[0]).to eq("foo.domain.com")
     end
 
     describe "and strict hostname checking is enabled" do
-      it "only uses the passed-in key" do
+      before do
         Puppet[:strict_hostname_checking] = true
+        @node = Puppet::Node.new("foo.domain.com",
```
Comment (mwaggett) [inline spec/unit/node_spec.rb]: same here

Comment (justinstoller) [reply spec/unit/node_spec.rb]: Good catch. I re-wrote the setup and teardown of these specs in an intermediate step and forgot they weren't needed when I cleaned them up.

Comment (justinstoller) [reply spec/unit/node_spec.rb]: pushed an update addressing your comments

**Comment by joshcooper** [comment]
I reverted the whitespace change in 5.5.x, so now there's a conflict. @justinstoller can you resolve?

**Comment by justinstoller** [comment]
Wow, I sure screwed up that rebase. I think it's sorted now.

**Comment by joshcooper** [comment]
I'll merge this up to 6.4.x before tonight's run
